proxy/context.go

```diff
@@ -233,10 +237,104 @@ func (c *context) clone() *context {
 	return &cc
 }
 
+func (c *context) wasExecuted()bool {
+	return c.executionCounter != 0
+}
+
 func (c *context) setMetricsPrefix(prefix string) {
 	c.metrics.prefix = prefix + ".custom."
 }
 
+func (tt *teeTie) Read(b []byte) (int, error) {
+	n, err := tt.r.Read(b)
+
+	if err != nil && err != io.EOF {
+		tt.w.CloseWithError(err)
+		return n, err
+	}
+
+	if n > 0 {
+		if _, werr := tt.w.Write(b[:n]); werr != nil {
+			log.Error("tee: error while tee request", werr)
+		}
+	}
+
+	if err == io.EOF {
+		tt.w.Close()
+	}
+
+	return n, err
+}
+
+func (tt *teeTie) Close() error { return nil }
+
+// Returns the cloned request and the tee body to be used on the main request.
+func cloneRequest(u *url.URL, req *http.Request) (*http.Request, io.ReadCloser, error) {
+	h := make(http.Header)
+	for k, v := range req.Header {
+		h[k] = v
+	}
+
+	var teeBody io.ReadCloser
+	mainBody := req.Body
+
+	// see proxy.go:231
+	if req.ContentLength != 0 {
```

Comment:
we cannot rely on ContentLength, because it is not always set, like in case of chunked content in the request. I think the teeTie body should work for 0 length body, too, so we could just skip the condition.

Reply:
my previous comment was wrong. See: https://github.com/zalando/skipper/blob/master/proxy/proxy.go#L488